File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        
        unordered_map<string, vector<string>> mp;
        for (const auto& str: strs){
            unordered_map<char, int> count;
            for (char c: str){
                count[c]++;
            }
            
            vector<pair<char, int>> keyval(count.begin(), count.end());
            sort(keyval.begin(), keyval.end());
            
            string key;
            for (auto& kv : keyval){
                key += kv.first + to_string(kv.second);
            }
            mp[key].push_back(str);
        }
        
        vector<vector<string>> res;
        for (auto& kv : mp){
            res.push_back(kv.second);
        }
        
        return res;
    }
};

File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        from collections import Counter, defaultdict
        mp = defaultdict(list)
        counter = []
        for s in strs:
            key_val = sorted(Counter(s).items(), key=lambda x: x[0])
            key = ''.join(str(k) + str(v) for k, v in key_val)
            mp[key].append(s)
        
        return [val for key, val in mp.items()]